File: mediaplayer_test/app/src/main/java/com/daeseong/mediaplayer_test/Main4Activity.java
package com.daeseong.mediaplayer_test;

import androidx.appcompat.app.AppCompatActivity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.graphics.Color;
import android.media.AudioManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageButton;
import android.widget.SeekBar;
import android.widget.TextView;

public class Main4Activity extends AppCompatActivity {

    private static final String TAG = Main4Activity.class.getSimpleName();

    private ImageButton btnPlay, btnPause, btnPrevious, btnNextgo, btnSearch;
    private TextView txtStartTime, txtEndTime;
    private SeekBar TimeBar, volumeBar;
    private AudioManager audioManager;
    private PlayerTimer playerTimer;
    private Mp3Player mp3Player;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initTitleBar();

        setContentView(R.layout.activity_main4);

        InitControl();

        initVolume();

        initPlaytime();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        try {
            mp3Player.release();
            stopPlayerTimer();
        } catch (Exception ex) {
            Log.d(TAG, ex.getMessage().toString());
        }
    }

    private void initTitleBar() {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(Color.rgb(255, 255, 255));
        }

        try {
            //안드로이드 8.0 오레오 버전에서만 오류 발생
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
            supportRequestWindowFeature(Window.FEATURE_NO_TITLE);
        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage().toString());
        }
    }

    private void initVolume() {
        audioManager = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
        int maxvolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
        int curvolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);

        volumeBar = findViewById(R.id.volumeBar);
        volumeBar.setMax(maxvolume);
        volumeBar.setProgress(curvolume);

        volumeBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, progress, 0);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    private void initPlaytime() {
        TimeBar = findViewById(R.id.TimeBar);
        TimeBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (!mp3Player.isPlaying()) {
                    mp3Player.seekTo(progress);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    private void initDisplaytime() {
        int nEndTime = mp3Player.getDuration() / 1000;
        int nEndMinutes = (nEndTime / 60) % 60;
        int nEndSeconds = nEndTime % 60;

        int nCurrentTime = mp3Player.getCurrentPosition() / 1000;
        int nCurrentMinutes = (nCurrentTime / 60) % 60;
        int nCurrentSeconds = nCurrentTime % 60;

        txtStartTime.setText(String.format("%02d:%02d", nCurrentMinutes, nCurrentSeconds));
        txtEndTime.setText(String.format("%02d:%02d", nEndMinutes, nEndSeconds));
    }

    private void InitControl() {
        mp3Player = Mp3Player.getInstance();

        txtStartTime = findViewById(R.id.startTime);
        txtEndTime = findViewById(R.id.endTime);

        btnSearch = findViewById(R.id.btnSearch);
        btnSearch.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    mp3Player.release();
                    mp3Player.play(Main4Activity.this, new Mp3Player.OnMediaPlayerListener() {
                        @Override
                        public void onCompletion(boolean bComplete) {
                            Log.e(TAG, "onCompletion");
                            stopPlayerTimer();
                        }

                        @Override
                        public void onPrepared(int mDuration) {
                            Log.e(TAG, "onPrepared");
                            setSeekBarProgress();
                        }
                    });

                    btnPlay.setVisibility(View.VISIBLE);
                    btnPause.setVisibility(View.INVISIBLE);
                } catch (Exception ex) {
                    Log.d(TAG, ex.getMessage().toString());
                }
            }
        });

        // 연주
        btnPlay = findViewById(R.id.btnPlay);
        btnPlay.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mp3Player.start();
                btnPlay.setVisibility(View.INVISIBLE);
                btnPause.setVisibility(View.VISIBLE);
            }
        });

        // 일시정지
        btnPause = findViewById(R.id.btnPause);
        btnPause.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mp3Player.isPlaying()) {
                    mp3Player.pause();
                    btnPlay.setVisibility(View.VISIBLE);
                    btnPause.setVisibility(View.INVISIBLE);
                }
            }
        });

        // 5초 뒤로
        btnPrevious = findViewById(R.id.btnPrevious);
        btnPrevious.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mp3Player.seekTo(mp3Player.getCurrentPosition() - 5000);
            }
        });

        // 5초 앞으로
        btnNextgo = findViewById(R.id.btnNextgo);
        btnNextgo.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mp3Player.seekTo(mp3Player.getCurrentPosition() + 5000);
            }
        });
    }

    private void releaseMediaPlayer() {
        try {
            if (mp3Player != null) {
                mp3Player.stop();
                mp3Player.release();
                mp3Player = null;
            }
        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage());
        }
    }

    private void stopPlayerTimer() {
        if (playerTimer != null) {
            playerTimer.stop();
            playerTimer = null;
        }
    }

    private void setSeekBarProgress() {
        stopPlayerTimer();

        playerTimer = new PlayerTimer();
        playerTimer.setCallback(new PlayerTimer.Callback() {
            @Override
            public void onTick(long timeMillis) {
                int position = mp3Player.getCurrentPosition();
                int duration = mp3Player.getDuration();

                if (duration <= 0) return;

                TimeBar.setMax(duration);
                TimeBar.setProgress(position);

                int nEndTime = duration / 1000;
                int nEndMinutes = (nEndTime / 60) % 60;
                int nEndSeconds = nEndTime % 60;

                int nCurrentTime = position / 1000;
                int nCurrentMinutes = (nCurrentTime / 60) % 60;
                int nCurrentSeconds = nCurrentTime % 60;

                txtStartTime.setText(String.format("%02d:%02d", nCurrentMinutes, nCurrentSeconds));
                txtEndTime.setText(String.format("%02d:%02d", nEndMinutes, nEndSeconds));
            }
        });
        playerTimer.start();
    }
}

File: mediaplayer_test(kotlin)/app/src/main/java/com/daeseong/mediaplayer_test/Main4Activity.kt
package com.daeseong.mediaplayer_test

import android.content.Context
import android.media.AudioManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.Window
import android.view.WindowInsets
import android.view.WindowManager
import android.widget.ImageButton
import android.widget.SeekBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class Main4Activity : AppCompatActivity() {

    private val tag = Main4Activity::class.java.simpleName

    private lateinit var btnPlay: ImageButton
    private lateinit var btnPause: ImageButton
    private lateinit var btnPrevious: ImageButton
    private lateinit var btnNextgo: ImageButton
    private lateinit var btnSearch: ImageButton
    private lateinit var txtStartTime: TextView
    private lateinit var txtEndTime: TextView
    private lateinit var TimeBar: SeekBar
    private lateinit var volumeBar: SeekBar
    private lateinit var audioManager: AudioManager

    private var mp3Player: Mp3Player? = null
    private var playerTimer: PlayerTimer? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initTitleBar()

        setContentView(R.layout.activity_main4)

        InitControl()

        initVolume()

        initPlaytime()
    }

    override fun onDestroy() {
        super.onDestroy()

        try {
            mp3Player?.release()
            stopPlayerTimer()
        } catch (ex: Exception) {
            Log.d(tag, ex.message.toString())
        }
    }

    private fun initTitleBar() {
        try {
            //안드로이드 8.0 오레오 버전에서만 오류 발생
            supportRequestWindowFeature(Window.FEATURE_NO_TITLE)
        } catch (ex: Exception) {
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.decorView.windowInsetsController?.hide(WindowInsets.Type.statusBars())
        } else {
            @Suppress("DEPRECATION")
            window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        }
    }

    private fun initVolume() {
        audioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        val currentVolume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC)

        volumeBar = findViewById(R.id.volumeBar)
        volumeBar.max = maxVolume
        volumeBar.progress = currentVolume

        volumeBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, progress, 0)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}

            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
    }

    private fun initPlaytime() {
        TimeBar = findViewById(R.id.TimeBar)
        TimeBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (!mp3Player?.isPlaying()!!) {
                    mp3Player?.seekTo(progress)
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}

            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
    }

    private fun initDisplaytime() {
        val nEndTime = mp3Player?.getDuration()?.div(1000)
        val nEndMinutes = (nEndTime?.div(60))?.rem(60)
        val nEndSeconds = nEndTime?.rem(60)

        val nCurrentTime = mp3Player?.getCurrentPosition()?.div(1000)
        val nCurrentMinutes = (nCurrentTime?.div(60))?.rem(60)
        val nCurrentSeconds = nCurrentTime?.rem(60)

        txtStartTime.text = String.format("%02d:%02d", nCurrentMinutes, nCurrentSeconds)
        txtEndTime.text = String.format("%02d:%02d", nEndMinutes, nEndSeconds)
    }

    private fun InitControl() {
        mp3Player = Mp3Player.getInstance()

        txtStartTime = findViewById(R.id.startTime)
        txtEndTime = findViewById(R.id.endTime)

        btnSearch = findViewById(R.id.btnSearch)
        btnSearch.setOnClickListener {
            try {
                mp3Player?.release()
                mp3Player?.play(this, object : Mp3Player.OnMediaPlayerListener {
                    override fun onCompletion(bComplete: Boolean) {
                        Log.e(tag, "onCompletion")
                        stopPlayerTimer()
                    }

                    override fun onPrepared(mDuration: Int) {
                        Log.e(tag, "onPrepared")
                        setSeekBarProgress()
                    }
                })

                btnPlay.visibility = View.VISIBLE
                btnPause.visibility = View.INVISIBLE
            } catch (ex: Exception) {
                Log.d(tag, ex.message.toString())
            }
        }

        // Play
        btnPlay = findViewById(R.id.btnPlay)
        btnPlay.setOnClickListener {
            mp3Player?.start()
            btnPlay.visibility = View.INVISIBLE
            btnPause.visibility = View.VISIBLE
        }

        // Pause
        btnPause = findViewById(R.id.btnPause)
        btnPause.setOnClickListener {
            if (mp3Player?.isPlaying() == true) {
                mp3Player?.pause()
                btnPlay.visibility = View.VISIBLE
                btnPause.visibility = View.INVISIBLE
            }
        }

        // 5 seconds backward
        btnPrevious = findViewById(R.id.btnPrevious)
        btnPrevious.setOnClickListener {
            mp3Player?.seekTo(mp3Player?.getCurrentPosition()!! - 5000)
        }

        // 5 seconds forward
        btnNextgo = findViewById(R.id.btnNextgo)
        btnNextgo.setOnClickListener {
            mp3Player?.seekTo(mp3Player?.getCurrentPosition()!! + 5000)
        }
    }

    private fun releaseMediaPlayer() {
        try {
            if (mp3Player?.isPlaying() == true) {
                mp3Player?.stop()
            }
            mp3Player?.release()
            mp3Player = Mp3Player.getInstance()
        } catch (ex: Exception) {
            Log.e(tag, ex.message ?: "")
        }
    }

    private fun stopPlayerTimer() {
        playerTimer?.stop()
        playerTimer = null
    }

    private fun setSeekBarProgress() {
        stopPlayerTimer()

        playerTimer = PlayerTimer()
        playerTimer?.setCallback(object : PlayerTimer.Callback {
            override fun onTick(timeMillis: Long) {
                val position = mp3Player?.getCurrentPosition()!!
                val duration = mp3Player?.getDuration()!!

                if (duration <= 0) return

                TimeBar.max = duration
                TimeBar.progress = position

                val nEndTime = duration / 1000
                val nEndMinutes = (nEndTime / 60) % 60
                val nEndSeconds = nEndTime % 60

                val nCurrentTime = position / 1000
                val nCurrentMinutes = (nCurrentTime / 60) % 60
                val nCurrentSeconds = nCurrentTime % 60

                txtStartTime.text = String.format("%02d:%02d", nCurrentMinutes, nCurrentSeconds)
                txtEndTime.text = String.format("%02d:%02d", nEndMinutes, nEndSeconds)
            }
        })
        playerTimer?.start()
    }
}
